### DISCUSSION TITLE
built-in language servers

### DISCUSSION THREAD

**Discussion by estin** [helix-editor/helix#10129] [Ideas] [answered]
Hi!

If we look at  https://github.com/helix-editor/helix/pull/2608 as the first attempt to implement new completion source - paths.

It's use same structures and logic as for language servers completions

And that why Helix can use language-server terms  to implement
 - new completions and snippets
 - spell checks
 - commands
 - etc
 
To implement this Idea needs to solve next major tasks
  - allow to start embed built-in language server with in-memory communication by channels
  - share docs with this language servers on read-only mode to not duplicate data and simplify synchronization (only helix-editor can make transactions)
  
And this logic may be integrated with plugin system in language servers terms too.
  
This idea isn't conflicting with Helix way?

**Comment by the-mikedavis** [comment, marked as answer]
This is reversed. LSP should fit into more generic Helix-specific types and systems rather than Helix re-using LSP types to implement other features. For example a diagnostic from a spell checker can fit into a core Diagnostic type. Custom snippets should share code with LSP snippets rather than custom snippets being implemented only in terms of LSP. Simulating language servers for features like spell checking is pure overhead and unnecessarily binds us to the LSP spec and its types.

**Reply by estin** [reply]
Yep. I see. Solid bounds of LSP spec can be blocker in future
